openedx/core/djangoapps/user_api/accounts/views.py

```diff
@@ -515,3 +573,110 @@ def partial_update(self, request):
             return Response(text_type(exc), status=status.HTTP_400_BAD_REQUEST)
         except Exception as exc:  # pylint: disable=broad-except
             return Response(text_type(exc), status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+    @request_requires_username
+    def post(self, request):
+        """
+        POST /api/user/v1/accounts/retire/
+
+        {
+            'username': 'user_to_retire'
+        }
+
+        Retires the user with the given username.  This includes
+        retiring this username, the associates email address, and
+        any other PII associated with this user.
+
+        Note that this implementation is the "merge patch" implementation proposed in
+        https://tools.ietf.org/html/rfc7396. The content_type must be "application/merge-patch+json" or
+        else an error response with status code 415 will be returned.
+        """
+        username = request.data['username']
+        if is_username_retired(username):
+            return Response(status=status.HTTP_404_NOT_FOUND)
+
+        user_model = get_user_model()
+        try:
+            user = user_model.objects.get(username=username)
+
+            with transaction.atomic():
+                self.clear_pii_from_userprofile(user)
+                Registration.objects.filter(user=user).delete()
+                self.delete_users_profile_images(user)
+                self.delete_users_country_cache(user)
+                self.retire_users_data_sharing_consent(username)
+                self.retire_sapsf_data_transmission(user)
+                self.retire_user_from_pending_enterprise_customer_user(user)
+                self.retire_entitlement_support_detail(user)
+                # TODO: Password Reset links - https://openedx.atlassian.net/browse/PLAT-2104
+                # TODO: Delete OAuth2 records - https://openedx.atlassian.net/browse/EDUCATOR-2703
+                user.first_name = ''
+                user.last_name = ''
+                user.is_active = False
+                user.username = get_retired_username_by_username(user.username)
```

Comment:
We should use the `retired_username` from the `UserRetirementStatus` here instead of re-hashing it. That way it will be consistent with the hash from when they originally put in the request if we update the salts. Should we be hashing the email here as well, or is that done elsewhere?

Reply:
The email hashing is done in `DeactivateLogoutView.post()`.

What do you think of modifying `get_retired_username_by_username()` to lookup existing `UserRetirementStatus` records for the already hashed username, and having it return that if it exists (otherwise, return a new hash)?  Or I guess I could wrap that up inside a new function...

Reply:
I think updating `get_retired_username_by_username` makes a lot sense. 👍